libraries/core_libs/consensus/src/dag/dag.cpp

```diff
@@ -452,6 +453,35 @@ void DagManager::addDagBlock(DagBlock const &blk, SharedTransactions &&trxs, boo
                << " pivot = " << frontier_.pivot << " tips: " << frontier_.tips;
 }
 
+void DagManager::sendNonFinalizedBlocks(std::unordered_set<blk_hash_t> &&blocks_hashes,
+                                        const std::shared_ptr<network::tarcap::TaraxaPeer> &peer,
+                                        uint64_t peer_period) {
+  // Sending of blocks is in DagManager class to use DagManager mutex_ which ensures that there is no race condition
+  // between syncing and gossiping dag blocks
```

Comment:
So how is mutex_ solving this race condition ? I sort of struggle to find the new synchronization mechnism that replaced mutex_for_sending_dag_blocks_. Checked sendBlock() method, which is related to the gossiping and I dont see anything that would solve this race condition there ?

Reply:
So blocks for gossiping are sent from DagManager::addDagBlock which is under the DagManager::mutex_, blocks for dag syncing is now also sent from DagManager::sendNonFinalizedBlocks which is under the same DagManager::mutex_, so when we get sync request until it is processed we will not be gossiping blocks. I spent a whole day thinking on this and this was the cleanest/safest way I could think of to solve the issue. I am open for discussion if there is a better way to do this

Reply:
hmm yeah I see. I forgot what is actually the problem in sending sync dag blocks and gossiping new dag blocks concurrently though ?